x/pairing/keeper/scores/score.go

```diff
@@ -0,0 +1,213 @@
+	for _, req := range diffSlot.Reqs {
+		reqName := req.GetName()
+		weight := strategy[reqName] // not checking if key is found because it's verified in init()
+
+		for _, score := range scores {
+			newScoreComp := req.Score(*score.Provider)
+			if newScoreComp == 0 {
+				err := fmt.Errorf("score component is zero. score component name: %s, provider address: %s", reqName, score.Provider.Address)
+				panic(err)
+			}
+			newScoreComp = uint64(math.Pow(float64(newScoreComp), float64(weight)))
+
+			// divide by previous score component (if exists) and multiply by new score
+			prevReqScoreComp, ok := score.ScoreComponents[reqName]
+			if ok {
+				score.Score /= prevReqScoreComp
+			}
+			score.Score *= newScoreComp
+
+			// update the score component map
+			score.ScoreComponents[reqName] = newScoreComp
+		}
+	}
+
+	return nil
+}
+
+// PrepareHashData prepares the hash needed in the pseudo-random choice of providers
+func PrepareHashData(projectIndex string, chainID string, epochHash []byte) []byte {
+	hashData := []byte{}
+
+	// add the session start block hash to the function to make it as unpredictable as we can
+	hashData = append(hashData, epochHash...)
+	hashData = append(hashData, chainID...)      // to make this pairing unique per chainID
+	hashData = append(hashData, projectIndex...) // to make this pairing unique per consumer
+
+	return hashData
+}
+
+// PickProviders pick a <group-count> providers set with a pseudo-random weighted choice (using the providers' score list and hashData)
+func PickProviders(ctx sdk.Context, scores []*PairingScore, groupCount uint64, hashData []byte, chosenProvidersIdx *map[int]bool) (returnedProviders []epochstoragetypes.StakeEntry) {
+	if len(scores) == 0 {
+		return returnedProviders
+	}
+
+	scoreSum := sdk.NewUint(0)
+	for _, providerScore := range scores {
+		scoreSum = scoreSum.Add(sdk.NewUint(providerScore.Score))
+	}
+	if scoreSum.IsZero() {
+		err := fmt.Errorf("score sum is zero. Cannot pick providers for pairing")
+		panic(err)
+	}
+
+	indexToSkip := *chosenProvidersIdx
+	for it := 0; it < int(groupCount); it++ {
+		hash := tendermintcrypto.Sha256(hashData) // TODO: we use cheaper algo for speed
+		bigIntNum := new(big.Int).SetBytes(hash)
+		hashAsNumber := sdk.NewUintFromBigInt(bigIntNum)
+		modRes := hashAsNumber.Mod(scoreSum)
+
+		newScoreSum := sdk.NewUint(0)
+
+		for idx := len(scores) - 1; idx >= 0; idx-- {
+			providerScore := scores[idx]
+			if indexToSkip[idx] {
+				// this is an index we added
+				continue
+			}
+			newScoreSum = newScoreSum.Add(sdk.NewUint(providerScore.Score))
+			if modRes.LT(newScoreSum) {
+				// we hit our chosen provider
+				returnedProviders = append(returnedProviders, *providerScore.Provider)
+				scoreSum = scoreSum.Sub(sdk.NewUint(providerScore.Score)) // we remove this provider from the random pool, so the sum is lower now
+				indexToSkip[idx] = true
+				break
+			}
+		}
+		if uint64(len(returnedProviders)) >= groupCount {
```

Comment:
make `groupcount` type be int and save the trouble of conversions

Reply:
done. see https://github.com/lavanet/lava/commit/2cabf1d176e04902c5eb6e41490706fe0783b514

Reply:
and I think this test is redundant?